### PR TITLE
hls: open standalone files

### PR DESCRIPTION
**Pull request by uninhm**

### CODE REVIEW

File: lua/lspconfig/hls.lua [hunk 1]
```diff
@@ -5,7 +5,11 @@ configs.hls = {
   default_config = {
     cmd = { 'haskell-language-server-wrapper', '--lsp' },
     filetypes = { 'haskell', 'lhaskell' },
-    root_dir = util.root_pattern('*.cabal', 'stack.yaml', 'cabal.project', 'package.yaml', 'hie.yaml'),
+    root_dir = function (fname)
+      local root_dir = util.root_pattern('*.cabal', 'stack.yaml', 'cabal.project', 'package.yaml', 'hie.yaml')(fname)
+      return root_dir
+        or vim.fn.getcwd()
```
Comment (lithammer) [inline lua/lspconfig/hls.lua:11]: You probably want to use `util.path.dirname(fname)` instead of `vim.fn.getcwd()`.

**Comment by mjlbach** [comment]
I'll be deprecating all util.path.dirname fallbacks (per server) soon, at which point you'll be able to default all servers to opening single files (if you choose). This is a bad idea though.

**Comment by uninhm** [comment]
> This is a bad idea though.

Why? I want to open a single file and have autocompletion...

**Comment by mjlbach** [comment]
> Why? I want to open a single file and have autocompletion...

Root directory is not something you can negotiate with the server. If you have a haskell "project" that consists of nested folders, and you open a nested haskell file (folderA -> myhaskellmod1.hs folderb->myhaskellmod2.hs) with util.path.dirname, the root directory for myhaskellmod1.hs will be folderA, HLS will not index myhaskellmod2.hs and autocompletion results will be completely broken (along with any imports).

**Comment by uninhm** [comment]
> If you have a haskell "project" that consists of nested folders...

Yeah, that's the problem, I don't have a haskell project, I have a single file, and I want to edit it...

**Comment by mjlbach** [comment]
You can edit it, just don't expect language servers to work, and you if you really want the above change add it locally in your config.

**Comment by uninhm** [comment]
> just don't expect language servers to work

Some of them work, so, why not to use them?

**Comment by oblitum** [comment]
@uninhm you're right (https://github.com/neovim/nvim-lspconfig/pull/1151#issuecomment-911023310). nvim lsp devs only recently are realizing a mistake reading the spec (https://github.com/neovim/neovim/pull/15430) while the "random  workspace root confs"-boat has sailed, now it needs to be undone. That's why current state is what it is, and it's even more sad given only late nvim lsp nigthly

**Comment by mjlbach** [comment]
> nvim lsp devs only recently are realizing a mistake reading the spec 

@oblitum, I'm going to kindly ask you to stop. Your actions have crossed into harassment, and your presence is not welcome. You have admitted not being an lspconfig user, and your attacks have demotivated me to work on lspconfig, neovim at large, and honestly open source in general. I am a volunteer, and this is my appeal to you to stop following me around and making my day worse.

I did not "make a mistake when reading the spec". I am not the original author of lspconfig, I did not implement sending the root by default, and users were merging faulty "roots" into the project well before I took over. Lspconfig is *only* meant to be used with projects in its current form, that is true. Users began adding in the cwd/parent dir as a "hack" to get local imports working, and it spiraled. I have taken steps to rectify that (I started writing an lspconfig replacement a long time ago, but lost interest as the current model fits my usecase).

**Comment by oblitum** [comment]
Is this supposed to mean kind?? You should simply have asked it, this is your first interaction with me here. I'm just passing on info for people, and they're just positively closing their PR, learning what's up, thumbs up my comments..... [I myself have stated already I should stop](https://github.com/neovim/nvim-lspconfig/pull/1203#issuecomment-911027358) due to the number of issues open on this matter, it's countless, but each of those PR is one different person, so, it's just repeatedly shared....

**Comment by mjlbach** [comment]
Please leave. This is my last warning.

**Pull request closed without merging**